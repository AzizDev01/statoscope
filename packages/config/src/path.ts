import path from 'path';
import module from 'module';

export function normalizePath(source: string, rootDir: string): string {
  return source.replace('<rootDir>', rootDir);
}

export function makeRequireFromPath(pathname: string): NodeRequire {
  return module.createRequire(path.resolve(pathname, '_'));
}

export function resolveAliasPackage(
  prefixes: string[],
  name: string,
  fromDir: string
): string {
  const localRequire = makeRequireFromPath(fromDir);
  name = normalizePath(name, fromDir);

  if (name.startsWith('.') || path.isAbsolute(name)) {
    localRequire(name);
    return name;
  }

  let packageNamespace: string;
  let packageName: string;

  if (name.startsWith('@')) {
    [, packageNamespace = '', packageName = ''] = name.match(/(@.+?)[\\/](.+)/) ?? [];
  } else {
    packageNamespace = '';
    packageName = name;
  }

  const paths = [
    [name],
    ...prefixes.map((prefix) => [packageNamespace, `${prefix}-${packageName}`]),
  ].map((item) => path.join(...(item.filter(Boolean) as string[])));

  for (const item of paths) {
    try {
      localRequire(item);
      return item;
      // eslint-disable-next-line no-empty
    } catch (e) {}
  }

  throw new Error(
    `Can't resolve ${name} with prefixes [${prefixes}].${
      !name.startsWith('.') ? 'Did you forget to install it?' : ''
    }`
  );
}
