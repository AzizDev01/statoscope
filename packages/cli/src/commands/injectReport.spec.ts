import fs from 'fs';
import path from 'path';
import { spawn } from 'child_process';

const rootPath = path.resolve(__dirname, '../../../../');
const outputDir = path.join(
  rootPath,
  'test/temp',
  Date.now().toString(),
  path.relative(rootPath, __filename)
);
const statsDir = path.resolve(
  __dirname,
  '../../../../test/fixtures/cli/injectReport/stats/'
);
const stats = [
  'stats-with-reports.json',
  'stats-without-extensions.json',
  'stats-without-meta.json',
  'stats-without-reports.json',
];

fs.mkdirSync(outputDir, { recursive: true });

jest.mock('fs', () => {
  const ofs = jest.requireActual('fs');
  return {
    ...ofs,
    // eslint-disable-next-line @typescript-eslint/explicit-function-return-type
    readFileSync(name: string, ...args: unknown[]): unknown {
      // eslint-disable-next-line eqeqeq
      if (name == '0') {
        return '';
      }
      return ofs.readFileSync(name, ...args);
    },
  };
});

const reports = [
  require.resolve(
    '../../../../test/fixtures/cli/injectReport/reports/single-report.json'
  ),
  require.resolve(
    '../../../../test/fixtures/cli/injectReport/reports/multiple-reports.json'
  ),
];

test.each(stats)('%s,', async (filename) => {
  for (const report of reports) {
    const proc = spawn(path.resolve(__dirname, '../../bin/cli.js'), [
      'inject-report',
      '--input',
      path.join(statsDir, filename),
      '--report',
      report,
    ]);

    const output = await new Promise((resolve, reject) => {
      const stdoutLines: string[] = [];
      const stderrLines: string[] = [];
      proc.stdout?.on('data', (data) => stdoutLines.push(data.toString()));
      proc.stderr?.on('data', (data) => stderrLines.push(data.toString()));
      proc.on('exit', (code) => {
        if (!code) {
          resolve(stdoutLines);
        } else {
          console.error(stderrLines);
          reject(new Error(`Exit code ${code}`));
        }
      });
    });

    expect(output).toMatchSnapshot();
  }
});
