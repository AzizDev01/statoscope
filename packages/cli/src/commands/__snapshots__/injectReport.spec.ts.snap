// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stats-with-reports.json, 1`] = `"{\\"__statoscope\\":{\\"extensions\\":[{\\"descriptor\\":{\\"name\\":\\"foo\\"}},{\\"descriptor\\":{\\"name\\":\\"@statoscope/stats-extension-custom-reports\\",\\"version\\":\\"5.13.1\\",\\"author\\":{\\"name\\":\\"Sergey Melyukov\\",\\"url\\":\\"https://twitter.com/smelukov\\"},\\"homepage\\":\\"https://github.com/statoscope/statoscope/\\",\\"description\\":\\"Statoscope extension to store custom reports in stats\\"},\\"payload\\":{\\"compilations\\":[{\\"id\\":null,\\"reports\\":[{\\"id\\":\\"baz\\",\\"view\\":[]},{\\"id\\":\\"foo\\",\\"data\\":[1,2,3],\\"view\\":[\\"struct\\"]}]}]}}]}}"`;

exports[`stats-with-reports.json, 2`] = `"{\\"__statoscope\\":{\\"extensions\\":[{\\"descriptor\\":{\\"name\\":\\"foo\\"}},{\\"descriptor\\":{\\"name\\":\\"@statoscope/stats-extension-custom-reports\\",\\"version\\":\\"5.13.1\\",\\"author\\":{\\"name\\":\\"Sergey Melyukov\\",\\"url\\":\\"https://twitter.com/smelukov\\"},\\"homepage\\":\\"https://github.com/statoscope/statoscope/\\",\\"description\\":\\"Statoscope extension to store custom reports in stats\\"},\\"payload\\":{\\"compilations\\":[{\\"id\\":null,\\"reports\\":[{\\"id\\":\\"baz\\",\\"view\\":[]},{\\"id\\":\\"foo\\",\\"data\\":[1,2,3],\\"view\\":[\\"struct\\"]},{\\"id\\":\\"bar\\",\\"data\\":[4,5,6],\\"view\\":[\\"list\\"]}]}]}}]}}"`;

exports[`stats-without-extensions.json, 1`] = `"{\\"__statoscope\\":{\\"extensions\\":[{\\"descriptor\\":{\\"name\\":\\"@statoscope/stats-extension-custom-reports\\",\\"version\\":\\"5.13.1\\",\\"author\\":{\\"name\\":\\"Sergey Melyukov\\",\\"url\\":\\"https://twitter.com/smelukov\\"},\\"homepage\\":\\"https://github.com/statoscope/statoscope/\\",\\"description\\":\\"Statoscope extension to store custom reports in stats\\"},\\"payload\\":{\\"compilations\\":[{\\"id\\":null,\\"reports\\":[{\\"id\\":\\"foo\\",\\"data\\":[1,2,3],\\"view\\":[\\"struct\\"]}]}]}}]}}"`;

exports[`stats-without-extensions.json, 2`] = `"{\\"__statoscope\\":{\\"extensions\\":[{\\"descriptor\\":{\\"name\\":\\"@statoscope/stats-extension-custom-reports\\",\\"version\\":\\"5.13.1\\",\\"author\\":{\\"name\\":\\"Sergey Melyukov\\",\\"url\\":\\"https://twitter.com/smelukov\\"},\\"homepage\\":\\"https://github.com/statoscope/statoscope/\\",\\"description\\":\\"Statoscope extension to store custom reports in stats\\"},\\"payload\\":{\\"compilations\\":[{\\"id\\":null,\\"reports\\":[{\\"id\\":\\"foo\\",\\"data\\":[1,2,3],\\"view\\":[\\"struct\\"]},{\\"id\\":\\"bar\\",\\"data\\":[4,5,6],\\"view\\":[\\"list\\"]}]}]}}]}}"`;

exports[`stats-without-meta.json, 1`] = `"{\\"__statoscope\\":{\\"extensions\\":[{\\"descriptor\\":{\\"name\\":\\"@statoscope/stats-extension-custom-reports\\",\\"version\\":\\"5.13.1\\",\\"author\\":{\\"name\\":\\"Sergey Melyukov\\",\\"url\\":\\"https://twitter.com/smelukov\\"},\\"homepage\\":\\"https://github.com/statoscope/statoscope/\\",\\"description\\":\\"Statoscope extension to store custom reports in stats\\"},\\"payload\\":{\\"compilations\\":[{\\"id\\":null,\\"reports\\":[{\\"id\\":\\"foo\\",\\"data\\":[1,2,3],\\"view\\":[\\"struct\\"]}]}]}}]}}"`;

exports[`stats-without-meta.json, 2`] = `"{\\"__statoscope\\":{\\"extensions\\":[{\\"descriptor\\":{\\"name\\":\\"@statoscope/stats-extension-custom-reports\\",\\"version\\":\\"5.13.1\\",\\"author\\":{\\"name\\":\\"Sergey Melyukov\\",\\"url\\":\\"https://twitter.com/smelukov\\"},\\"homepage\\":\\"https://github.com/statoscope/statoscope/\\",\\"description\\":\\"Statoscope extension to store custom reports in stats\\"},\\"payload\\":{\\"compilations\\":[{\\"id\\":null,\\"reports\\":[{\\"id\\":\\"foo\\",\\"data\\":[1,2,3],\\"view\\":[\\"struct\\"]},{\\"id\\":\\"bar\\",\\"data\\":[4,5,6],\\"view\\":[\\"list\\"]}]}]}}]}}"`;

exports[`stats-without-reports.json, 1`] = `"{\\"__statoscope\\":{\\"extensions\\":[{\\"descriptor\\":{\\"name\\":\\"foo\\"}},{\\"descriptor\\":{\\"name\\":\\"@statoscope/stats-extension-custom-reports\\",\\"version\\":\\"5.13.1\\",\\"author\\":{\\"name\\":\\"Sergey Melyukov\\",\\"url\\":\\"https://twitter.com/smelukov\\"},\\"homepage\\":\\"https://github.com/statoscope/statoscope/\\",\\"description\\":\\"Statoscope extension to store custom reports in stats\\"},\\"payload\\":{\\"compilations\\":[{\\"id\\":null,\\"reports\\":[{\\"id\\":\\"foo\\",\\"data\\":[1,2,3],\\"view\\":[\\"struct\\"]}]}]}}]}}"`;

exports[`stats-without-reports.json, 2`] = `"{\\"__statoscope\\":{\\"extensions\\":[{\\"descriptor\\":{\\"name\\":\\"foo\\"}},{\\"descriptor\\":{\\"name\\":\\"@statoscope/stats-extension-custom-reports\\",\\"version\\":\\"5.13.1\\",\\"author\\":{\\"name\\":\\"Sergey Melyukov\\",\\"url\\":\\"https://twitter.com/smelukov\\"},\\"homepage\\":\\"https://github.com/statoscope/statoscope/\\",\\"description\\":\\"Statoscope extension to store custom reports in stats\\"},\\"payload\\":{\\"compilations\\":[{\\"id\\":null,\\"reports\\":[{\\"id\\":\\"foo\\",\\"data\\":[1,2,3],\\"view\\":[\\"struct\\"]},{\\"id\\":\\"bar\\",\\"data\\":[4,5,6],\\"view\\":[\\"list\\"]}]}]}}]}}"`;
